packages/spacecat-shared-rum-api-client/src/functions/formVitals.js

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2024 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+const FORM_SOURCE = ['.form', '.marketo', '.marketo-form'];
+const BOT = 'bot';
+const CHECKPOINT_MAPPING = { formsubmit: 'formsubmit', viewblock: 'formview', click: 'formengagement' };
+
+function initializeResult(url, pageViews) {
+  return {
+    url,
+    formsubmit: {},
+    formview: {},
+    formengagement: {},
+    pageview: pageViews[url],
+  };
+}
+
+function collectFormVitals(bundles, pageViews) {
+  const results = {};
+  for (const bundle of bundles) {
+    const {
+      url, userAgent, weight, events,
+    } = bundle;
+
+    if (userAgent && !userAgent.startsWith(BOT)) {
+      // Track if each condition has been processed for this event
+      const processedCheckpoints = {
+        viewblock: false,
+        formsubmit: false,
+        click: false,
+      };
+
+      // Process each event within the bundle
+      for (const event of events) {
+        const { checkpoint, source } = event;
+
+        // Only process the checkpoint once per event
+        if (!processedCheckpoints[checkpoint]) {
+          if ((checkpoint === 'viewblock' && FORM_SOURCE.includes(source)) || (checkpoint === 'formsubmit') || (checkpoint === 'click' && source && /\bform\b/.test(source.toLowerCase()))) {
+            results[url] = results[url] || initializeResult(url, pageViews);
+            const key = CHECKPOINT_MAPPING[checkpoint];
+            const res = results[url];
+            res[key] = {
+              ...res[key],
+              [userAgent]: (res[key][userAgent] || 0) + weight,
+            };
+            // Mark this checkpoint as processed
+            processedCheckpoints[checkpoint] = true;
+          }
+        }
+      }
+    }
+  }
+  return results;
+}
+
+function pageviewsByUrlAndUserAgent(bundles) {
+  return bundles.reduce((acc, cur) => {
+    const { userAgent } = cur;
+    if (!userAgent || userAgent.startsWith(BOT)) return acc;
+    acc[cur.url] = acc[cur.url] || {};
+    acc[cur.url][userAgent] = (acc[cur.url][userAgent] || 0) + cur.weight;
+    return acc;
+  }, {});
+}
+
+function handler(bundles) {
+  const pageViews = pageviewsByUrlAndUserAgent(bundles);
+  const formVitals = collectFormVitals(bundles, pageViews);
+  return Object.values(formVitals);
+}
+
+export default {
+  handler,
+  checkpoints: ['viewblock', 'formsubmit', 'click', 'error', 'top'],
```

Comment:
are you planning to use `'error', 'top'` checkpoints as well?

Reply:
@rpapani  **'top'** checkpoint we are using to count the pageviews for a particular url. We are counting pageviews by top checkpoint and form views by the three conditions mentioned in form-vitals.js
And **'error'** checkpoint will be used in next phase to get forms which has errors.

Reply:
> 'top' checkpoint we are using to count the pageviews for a particular url.

you don't need the `top` checkpoint for the pageviews. Even if you don't add `top` checkpoint, all the bundles will be still there, which will be used to calculate the pageviews. we'll only filter the events by checkpoints.

> And 'error' checkpoint will be used in next phase to get forms which has errors.

May be you can add it later, when you add the functionality.

Reply:
done.